localStorage shim, fixes #5195

zslayton opened the pull request:
@jdm This initial version has a few outstanding issues that I wanted to invite input on. Specifically:
1. I had some difficulty finding a home for the `StorageType` enum. Structs defined outside of the `script` module don't seem to be able to use the `#[jstraceable]` annotation and the `net` module (where `StorageTask` lives) doesn't have access to `script`. Per Simon Sapin's suggestion, I worked around this temporarily by creating a `TraceableStorageType` stand-in struct that was traceable and which could be translated into a regular `StorageType` when being sent to the `StorageTask`. Unsure of the best way to resolve this hack. Thoughts?
2. Apart from the `Storage` constructor used in `Window::SessionStorage` and the new `Window::LocalStorage`, there's also a method called `Storage::Constructor`. I'm unclear on what (if anything) will actually invoke this, so I'm not sure which variant of `StorageType` to use here. I've temporarily created an `Unknown` variant of `StorageType` as a placeholder.
3. I discovered that the web platform tests directory's localStorage tests. Many of them now pass despite the configured expectation that they fail. However, several do not pass. Is there a good way for me to add debug logging or otherwise get a sense of which assertion failed / what went wrong?

Thanks for your continued help!

zslayton committed:
localStorage shim, fixes #5195

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @jdm (or someone else) soon.

jdm commented:
A couple quick notes, since I don't have time for a proper review right now - the no_jsmanaged_fields! macro is what you're looking for instead of a separate type, and the Storage::Constructor method can be deleted - it looks like it was cargo-culted, but nobody calls it and the WebIDL declaration doesn't include a constructor. As for tests, you can see the particular assertion that failed in the output and correlate that to blocks in the test files that include the message.

zslayton commented:
Perfect, thanks! I'll try to push some revisions tomorrow.

On Sat, Mar 14, 2015 at 5:47 PM, Josh Matthews <notifications@github.com>
wrote:

> A couple quick notes, since I don't have time for a proper review right
> now - the no_jsmanaged_fields! macro is what you're looking for instead of
> a separate type, and the Storage::Constructor method can be deleted - it
> looks like it was cargo-culted, but nobody calls it and the WebIDL
> declaration doesn't include a constructor. As for tests, you can see the
> particular assertion that failed in the output and correlate that to blocks
> in the test files that include the message.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/servo/servo/pull/5217#issuecomment-80733849>.
>

zslayton commented:
I've removed all of the stopgap code I described in the original comment. I'm still sifting through the wpt test results to see what I need to mark as no longer expecting failure and check for regressions.

zslayton commented:
I've traced the failed tests to what appears to be an inconsistency in how data types are handled when using direct property manipulation on the Storage object (`localStorage["foo"] = 7`) vs the functional APIs (`localStorage.setItem("foo", 7)`).

While values written and read with `setItem` and `getItem` are always stringified, those set and read as properties preserve their original data types. Since the WPT tests are looking for strings, this registers as a failure.

Example JS:
```javascript
localStorage.setItem("mouse", 7);
alert(typeof(localStorage.getItem("mouse")) + '/' + localStorage.getItem("mouse"));
alert(typeof(localStorage["mouse"]) + '/' + localStorage["mouse"]);
localStorage["cheese"] = 7;
alert(typeof(localStorage.getItem("cheese")) + '/' + localStorage.getItem("cheese"));
alert(typeof(localStorage["cheese"]) + '/' + localStorage["cheese"]);
```
Output:
```
 0:03.18 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:27668) "ALERT: string/7"
 0:03.18 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:27668) "ALERT: string/7"
 0:03.18 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:27668) "ALERT: string/7"
 0:03.18 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:27668) "ALERT: number/7"
```
Curiously, the analogous test file for sessionStorage passes without issue. I'm uncertain why `localStorage` would behave differently from `sessionStorage` given that they share the same API and much of the same implementation. Continuing to investigate.

Ms2ger commented:
That's a codegen bug. Filed as #5223.

zslayton commented:
Thanks, @Ms2ger! I'll run through the WPT .ini files and adjust the expectations for the tests. I'll document which ones are failing due to #5223 in a separate issue so someone can follow up once that's resolved.

zslayton commented:
My most recent commit removed any `expected: FAIL` configurations for tests which now pass. I've opened #5225 to track the new expected failures.

jdm commented:
Great job! I've left comments on Critic; one little style nit, and a bunch of empty test expectation files to delete.

zslayton commented:
@jdm Fixed!

jdm commented:
Excellent! Go ahead and squash all of these commits together!

zslayton commented:
@jdm Squashed!

jdm commented:
Thanks!

jdm commented:
```
/html/dom/interfaces.html
-------------------------
PASS expected FAIL Window interface: window must inherit property "localStorage" with the proper type (124)
/webstorage/event_local_key.html
--------------------------------
TIMEOUT [Parent]
/webstorage/event_local_newvalue.html
-------------------------------------
TIMEOUT [Parent]
/webstorage/event_local_oldvalue.html
-------------------------------------
TIMEOUT [Parent]
/webstorage/event_local_storagearea.html
----------------------------------------
TIMEOUT [Parent]
/webstorage/event_local_url.html
--------------------------------
TIMEOUT [Parent]
/webstorage/storage_local_builtins.html
---------------------------------------
OK expected TIMEOUT [Parent]
```

zslayton commented:
Shoot, I'm not sure how I missed those. I've updated the expectations and pushed. Since this wasn't part of the code review I went ahead and squashed the commits. I hope that's okay.

jdm commented:
Yep, that's perfectly fine.
